hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/BatchEntityInsideEmbeddableSelectFetchInitializer.java

```diff
@@ -0,0 +1,114 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.engine.spi.EntityKey;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.internal.log.LoggingHelper;
+import org.hibernate.metamodel.mapping.internal.ToOneAttributeMapping;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.spi.NavigablePath;
+import org.hibernate.sql.exec.spi.ExecutionContext;
+import org.hibernate.sql.results.graph.DomainResultAssembler;
+import org.hibernate.sql.results.graph.FetchParentAccess;
+import org.hibernate.sql.results.graph.entity.EntityInitializer;
+
+public class BatchEntityInsideEmbeddableSelectFetchInitializer extends AbstractBatchEntitySelectFetchInitializer {
+	private static final String CONCRETE_NAME = BatchEntityInsideEmbeddableSelectFetchInitializer.class.getSimpleName();
+
+	private Map<EntityKey, List<Object[]>> toBatchLoad = new HashMap<>();
+
+	public BatchEntityInsideEmbeddableSelectFetchInitializer(
+			FetchParentAccess parentAccess,
+			ToOneAttributeMapping referencedModelPart,
+			NavigablePath fetchedNavigable,
+			EntityPersister concreteDescriptor,
+			DomainResultAssembler identifierAssembler) {
+		super( parentAccess, referencedModelPart, fetchedNavigable, concreteDescriptor, identifierAssembler );
+	}
+
+	@Override
+	protected String getConcreteName() {
+		return CONCRETE_NAME;
+	}
+
+	@Override
+	protected void addParentInfo() {
+		final List<Object[]> parents = getBatchInfos();
+		parentAccess.registerResolutionListener(
+				o ->
+						parents.add(
+								new Object[] {
+										parentAccess.findFirstEntityInitilaizer().getEntityKey(),
+										o
+								}
+						)
+		);
+	}
+
+	private List<Object[]> getBatchInfos() {
+		List<Object[]> objects = toBatchLoad.get( entityKey );
+		if ( objects == null ) {
+			objects = new ArrayList<>();
+			toBatchLoad.put( entityKey, objects );
+		}
+		return objects;
+	}
+
+	@Override
+	public void endLoading(ExecutionContext context) {
+		final EntityInitializer entityInitializer = parentAccess.findFirstEntityInitilaizer();
+		final String propertyName = getPropertyName();
+		final int propertyIndex = getPropertyIndex( parentAccess, propertyName );
+		toBatchLoad.forEach(
+				(entityKey, parentInfos) -> {
+					final SharedSessionContractImplementor session = context.getSession();
+					final Object instance = loadInstance( entityKey, referencedModelPart, session );
+					for ( Object[] parentInfo : parentInfos ) {
+						final PersistenceContext persistenceContext = session.getPersistenceContext();
+						setInstance(
+								entityInitializer,
+								referencedModelPart,
+								propertyName,
+								propertyIndex,
+								session,
+								instance,
+								parentInfo[1],
+								persistenceContext.getEntry( persistenceContext.getEntity( (EntityKey) parentInfo[0] ) )
+						);
+					}
+				}
+		);
+		toBatchLoad = null;
```

Comment:
```suggestion
		toBatchLoad.clear();
```

to avoid a possible NPE

Reply:
:+1: